CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr

aocsa opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59012 CUDA support in the CSR layout:  CUDA addmm/matvec
* **#59011 CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr**

Restack of #57275, see review history there.

Differential Revision: [D28719550](https://our.internmc.facebook.com/intern/diff/D28719550)

aocsa committed:
CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr

[ghstack-poisoned]

bhosmer commented:
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28719550).

bhosmer reviewed (approved):
Transferring approval from https://github.com/pytorch/pytorch/pull/57275

ngimel commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -0,0 +1,160 @@
+                    cast_value,
+                    out_strides1
+                    ]__device__(int64_t irow) {
+                        int32_t start_index = crow_indices_accessor[irow];
```
How are you sure that the value fits in int32?

aocsa replied:
Sure, thanks for notice this It should be `index_t` dtype.

ngimel commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -0,0 +1,160 @@
+
+                index_t* crow_indices_accessor = src_crow_indices.data_ptr<index_t>();
+                index_t* col_indices_accessor = src_col_indices.data_ptr<index_t>();
+                int64_t out_storage_offset = resultBuffer.storage_offset();
```
why is it storage_offset here instead of resultBuffer.data_ptr<scalar_t>?

aocsa replied:
The storage_offset is used to compute the new `index` 

https://github.com/pytorch/pytorch/blob/a7b796c73bd02bc4fe26686d067c48fdd847e679/aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu#L132

And then is used with  `scalar_t* out_ptr = resultBuffer.data_ptr<scalar_t>();` to update the value https://github.com/pytorch/pytorch/blob/a7b796c73bd02bc4fe26686d067c48fdd847e679/aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu#L133

ngimel replied:
It doesn't look right. storage_offset (as the name suggests) gives offset of the beginning of the tensor withing the storage. In most cases it's 0. When you are computing index to write to, you should only be using indices and strides, not storage offset.

aocsa replied:
Thanks @ngimel, I will file an issue about this.

ngimel commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -0,0 +1,160 @@
+                auto policy = thrust::cuda::par(allocator).on(stream);
+
+               // Note that this could be wildly imbalanced if the sparsity pattern varies a lot between rows.
+               thrust::for_each(
```
I'm not sure how thrust::for_each works, but I suspect it parallelizes between threads, which means that data access pattern in the loop below is pretty bad

aocsa replied:
Sure, a custom CUDA kernel can be created like `sparseElementwiseKernel` however I suspect that even with a custom kernel the data access pattern will be still the same like here `sparseElementwiseKernel`. https://github.com/pytorch/pytorch/blob/9e60c7dee36f5a4fc3a119d364ab072dac2ce58a/aten/src/ATen/native/sparse/cuda/SparseCUDAApplyUtils.cuh#L66

ngimel replied:
In the kernel you linked neighboring threads are accessing neighboring values, not a single thread loops over all elements in this row. We can leave this as further performance improvement.

aocsa replied:
Sure I will file an issue to this too.

aocsa commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -0,0 +1,160 @@
+      output.is_contiguous(),
+      "out argument must be contiguous, but got: ",
+      output.suggest_memory_format());
+  TORCH_CHECK(
```
Note: check are automatically generated but is not always called. I think this should be addressed in a different PR as it not only related to CSR tensors but COO tensors as well.
I detailed the issue here: #59058

aocsa commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -0,0 +1,160 @@
+  Tensor src_crow_indices = src.crow_indices();
+  Tensor src_col_indices = src.col_indices();
+
+  resize_output(output, dense.sizes());
```
Not to the best of my knowledge. This function implements `dense` + `sparse_csr` addition operation and currently CSR is restricted to 2-dim.  Future n-dim support is listed as TODO here https://github.com/pytorch/pytorch/issues/44634 as GCS. 

Addition of sparse CSR tensors is in the list too and detailed here: https://github.com/pytorch/pytorch/issues/59060

ngimel replied:
Yeah, but `dense + sparse_csr` could support broadcasting as long as dense and sparse_csr sizes are compatible, it doesn't need GCS support for that.

aocsa replied:
Yes, COO proceeds in a similar way. 
https://github.com/pytorch/pytorch/blob/ba694520e5004b74b575614f9d7f86a26436d61b/aten/src/ATen/native/sparse/SparseTensorMath.cpp#L676

aocsa commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -0,0 +1,160 @@
+      commonDtype,
+      "add_out_op2_sparse_csr",
+      [&valuesBuffer, &resultBuffer, &alpha, &src_crow_indices, &src_col_indices]() {
+        AT_DISPATCH_INDEX_TYPES(
```
It's because operations like matvec and addmm, the CPU implementation relies on MKL, currently the APIs used here only works with int32 indices and when MKL is not enabled custom kernels for any 32 or 64 bits indices are used. This is better detailed here https://pearu.github.io/csr_tensor_support.html

aocsa commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -0,0 +1,160 @@
+      src.sizes(),
+      " (FYI: op2-sparse addition does not currently support broadcasting)");
+
+  auto commonDtype = promoteTypes(dense.scalar_type(), src.scalar_type());
```
If so, I think this should be addressed in a different PR as it is not only related to CSR but COO as well.  
https://github.com/pytorch/pytorch/blob/9e60c7dee36f5a4fc3a119d364ab072dac2ce58a/aten/src/ATen/native/sparse/SparseTensorMath.cpp#L673

ngimel replied:
Given that broadcast is not supported, tensor - scalar type promotion issues won't arise, so this is fine for now.

aocsa committed:
minor update/feedback comments addressed on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"


Restack of #57275, see review history there.

Differential Revision: [D28719550](https://our.internmc.facebook.com/intern/diff/D28719550)

[ghstack-poisoned]

aocsa commented:
Thanks for the feedback @ngimel, I've addressed the comments fixing where possible or giving referenced information. When issues are not only related to CSR I create new issues as detailed here:

Issue: support auto generation of device check for sparse tensors [https://github.com/pytorch/pytorch/issues/59058](https://github.com/pytorch/pytorch/issues/59058)
Issue: Add 64-bit indices support to csrmm2 [https://github.com/pytorch/pytorch/issues/58899](https://github.com/pytorch/pytorch/issues/58899)  

Hope there is not more new issues, If so unfortunately, my timing is tight,  as I won't be able to work on this after this month.  So make big changes will not be possible and I would prefer to create new issues and address them in future next PRs or figure out a handoff strategy. Thanks cc @bhosmer

bhosmer commented:
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28719550).
